src/mlpack/methods/decision_tree/decision_tree_regressor.hpp

```diff
@@ -0,0 +1,527 @@
+/**
+ * @file methods/decision_tree/decision_tree_regressor.hpp
+ * @author Rishabh Garg
+ *
+ * The decision tree regressor class. Its behavior can be controlled via the
+ * template arguments.
+ *
+ * mlpack is free software; you may redistribute it and/or modify it under the
+ * terms of the 3-clause BSD license.  You should have received a copy of the
+ * 3-clause BSD license along with mlpack.  If not, see
+ * http://www.opensource.org/licenses/BSD-3-Clause for more information.
+ */
+#ifndef MLPACK_METHODS_DECISION_TREE_DECISION_TREE_REGRESSOR_HPP
+#define MLPACK_METHODS_DECISION_TREE_DECISION_TREE_REGRESSOR_HPP
+
+#include <mlpack/prereqs.hpp>
+#include "mad_gain.hpp"
+#include "mse_gain.hpp"
+#include "best_binary_numeric_split.hpp"
+#include "all_categorical_split.hpp"
+#include "random_binary_numeric_split.hpp"
+#include "all_dimension_select.hpp"
+#include <type_traits>
+
+
+namespace mlpack {
+namespace tree {
+
+/**
+ * This class implements a generic decision tree learner.  Its behavior can be
+ * controlled via its template arguments.
+ *
+ * The class inherits from the auxiliary split information in order to prevent
+ * an empty auxiliary split information struct from taking any extra size.
+ */
+template<typename FitnessFunction = MSEGain,
+         template<typename> class NumericSplitType = BestBinaryNumericSplit,
+         template<typename> class CategoricalSplitType = AllCategoricalSplit,
+         typename DimensionSelectionType = AllDimensionSelect,
+         bool NoRecursion = false>
+class DecisionTreeRegressor :
+    public NumericSplitType<FitnessFunction>::AuxiliarySplitInfo,
+    public CategoricalSplitType<FitnessFunction>::AuxiliarySplitInfo
```

Comment:
Yeah sure. I will open an issue for both of them once this one gets merged. I don't want to open an issue right now because it creates a bit of confusion for someone and it also needs to be stated that this is WIP etc etc.

Don't worry, I keep a log of all details like this so that I don't forget them. And I have noted these down :)

Reply:
Sounds good!  I just want to make sure it doesn't fall off of the list. :)